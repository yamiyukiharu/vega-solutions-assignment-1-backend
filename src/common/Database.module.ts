import { Module } from "@nestjs/common";
import { TypeOrmModule } from "@nestjs/typeorm";
import { AppConfigModule } from "./AppConfig.module";
import { ConfigService } from "@nestjs/config";
import { Transaction } from "src/transaction/models/Transaction.model";

@Module({
  imports: [
    TypeOrmModule.forRootAsync({
      imports: [AppConfigModule],
      useFactory: (configService: ConfigService) => ({
        type: 'mongodb',
        url: configService.get('DB_URL'),
        entities: [Transaction],
        synchronize: true,
        useNewUrlParser: true,
        logging: true,
      }),
      inject: [ConfigService],
    }),
  ],
  exports: [TypeOrmModule]
})
export class DatabaseModule { }