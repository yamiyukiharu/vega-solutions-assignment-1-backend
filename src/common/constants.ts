export const REPORTS_QUEUE = 'transactions-report';

export const RECORD_QUEUE = 'transactions-record';