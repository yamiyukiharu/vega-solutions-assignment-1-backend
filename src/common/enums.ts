export enum Protocol {
  UNISWAPV3 = 'uniswapv3',
}

export enum Pool {
  ETH_USDC = 'usdc_eth',
}

export enum ReportStatus {
  PENDING = 'pending',
  IN_PROGRESS = 'in_progress',
  COMPLETED = 'completed',
  FAILED = 'failed',
}