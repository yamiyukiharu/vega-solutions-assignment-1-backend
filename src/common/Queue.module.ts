
import { BullModule } from '@nestjs/bull';
import { Module } from '@nestjs/common';
import { ConfigService } from '@nestjs/config';
import { AppConfigModule } from './AppConfig.module';

@Module({
  imports: [
    BullModule.registerQueueAsync({
      imports: [AppConfigModule],
      name: 'transactions-report',
      useFactory: async (configService: ConfigService) => ({
        redis: {
          host: configService.get('QUEUE_HOST'),
          port: configService.get('QUEUE_PORT'),
        },
      }),
      inject: [ConfigService],
    }),
  ],
  exports: [BullModule],
})
export class QueueModule { }
