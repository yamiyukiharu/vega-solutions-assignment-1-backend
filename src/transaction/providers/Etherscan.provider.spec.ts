import { Test, TestingModule } from '@nestjs/testing';
import { HttpService } from '@nestjs/axios';
import { ConfigService } from '@nestjs/config';
import { EtherscanProvider, EtherscanResponse } from './Etherscan.provider';
import { Pool } from 'src/common/enums';
import BigNumber from 'bignumber.js';

describe('EtherscanProvider', () => {
  let provider: EtherscanProvider;
  let httpService: HttpService;
  let configServiceGet = jest.fn();
  const mockAxiosGet = jest.fn();

  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      providers: [
        EtherscanProvider,
        {
          provide: HttpService,
          useValue: {
            axiosRef: {
              get: mockAxiosGet,
            },
          },
        },
        {
          provide: ConfigService,
          useValue: {
            get: configServiceGet,
          },
        },
      ],
    }).compile();

    provider = module.get<EtherscanProvider>(EtherscanProvider);
    httpService = module.get<HttpService>(HttpService);
    jest.resetAllMocks();
    configServiceGet.mockReturnValue('test-api-key');
  });

  describe('getTransactions', () => {
    const mockResponse: EtherscanResponse = {
      result: [
        {
          blockNumber: '1',
          timeStamp: '1630500000',
          hash: '0x123456789',
          blockHash: '0xabcdef',
          value: '1000000000000000000',
          gasPrice: '1000000000',
          gasUsed: '21000',
        },
      ],
    };

    it('should return an array of transactions', async () => {
      mockAxiosGet.mockResolvedValueOnce({ data: mockResponse });

      const result = await provider.getTransactions({
        pool: Pool.ETH_USDC,
        page: 0,
        limit: 100,
        startBlock: 0,
        endBlock: 27025780,
        sort: 'asc',
      });

      expect(result).toEqual([
        {
          id: '0x123456789',
          timestamp: 1630500000,
          blockNumber: 1,
          fee: '21000000000000',
        },
      ]);
    });

    it('should throw an error if pool is not supported', async () => {
      await expect(
        provider.getTransactions({
          pool: 'invalid-pool' as Pool,
          page: 0,
          limit: 100,
          startBlock: 0,
          endBlock: 27025780,
          sort: 'asc',
        }),
      ).rejects.toThrow('Pool not supported');
    });

    it('should use the correct contract address for ETH_USDC pool', async () => {
      const getPoolAddressSpy = jest
        .spyOn(provider, 'getPoolAddress')
        .mockReturnValueOnce('0x123');
      mockAxiosGet.mockResolvedValueOnce({ data: mockResponse });

      await provider.getTransactions({
        pool: Pool.ETH_USDC,
        page: 0,
        limit: 100,
        startBlock: 0,
        endBlock: 27025780,
        sort: 'asc',
      });

      expect(getPoolAddressSpy).toHaveBeenCalledWith(Pool.ETH_USDC);
      expect(mockAxiosGet.mock.calls[0]).toMatchSnapshot();
    });

    it('should use the correct API key', async () => {
      mockAxiosGet.mockResolvedValueOnce({ data: mockResponse });

      await provider.getTransactions({
        pool: Pool.ETH_USDC,
        page: 0,
        limit: 100,
        startBlock: 0,
        endBlock: 27025780,
        sort: 'asc',
      });

      expect(mockAxiosGet.mock.calls[0]).toMatchSnapshot();
    });

    it('should calculate the correct fee', async () => {
      mockAxiosGet.mockResolvedValueOnce({ data: mockResponse });

      const result = await provider.getTransactions({
        pool: Pool.ETH_USDC,
        page: 0,
        limit: 100,
        startBlock: 0,
        endBlock: 27025780,
        sort: 'asc',
      });

      expect(result[0].fee).toEqual(new BigNumber('21000000000000').toString());
    });
  });
});
