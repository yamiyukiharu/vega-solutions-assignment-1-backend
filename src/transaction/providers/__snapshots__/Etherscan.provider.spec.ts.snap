// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EtherscanProvider getTransactions should use the correct API key 1`] = `
[
  "https://api.etherscan.io/api?module=account&action=tokentx&address=0x88e6a0c2ddd26feeb64f039a2c41296fcb3f5640&contractAddress=0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2&page=0&offset=100&startblock=0&endblock=27025780&apiKey=test-api-key&sort=asc",
]
`;

exports[`EtherscanProvider getTransactions should use the correct contract address for ETH_USDC pool 1`] = `
[
  "https://api.etherscan.io/api?module=account&action=tokentx&address=0x123&contractAddress=0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2&page=0&offset=100&startblock=0&endblock=27025780&apiKey=test-api-key&sort=asc",
]
`;
